### PR TITLE
Add decision tree classifier factory

### PR DESCRIPTION
**Pull request by gyrdym**

### CODE REVIEW

File: lib/src/classifier/decision_tree_classifier/decision_tree_classifier.dart [hunk 1]
```diff
@@ -36,19 +35,16 @@ abstract class DecisionTreeClassifier implements Classifier, Assessable {
...
-      targetName,
-      minError,
-      minSamplesCount,
-      maxDepth,
-    );
-
-    return DecisionTreeClassifierImpl(solver, targetName, dtype);
-  }
+  }) => createDecisionTreeClassifier(
+    fittingData,
```
Comment (gyrdym) [inline lib/src/classifier/decision_tree_classifier/decision_tree_classifier.dart]: rename `fittingData` to `trainData`

File: lib/src/classifier/decision_tree_classifier/decision_tree_classifier_factory_impl.dart [hunk 2]
```diff
@@ -0,0 +1,19 @@
+import 'package:ml_algo/src/classifier/decision_tree_classifier/decision_tree_classifier.dart';
+import 'package:ml_algo/src/classifier/decision_tree_classifier/decision_tree_classifier_factory.dart';
+import 'package:ml_algo/src/tree_solver/tree_solver.dart';
+import 'package:ml_linalg/dtype.dart';
+
+import 'decision_tree_classifier_impl.dart';
```
Comment (gyrdym) [inline lib/src/classifier/decision_tree_classifier/decision_tree_classifier_factory_impl.dart]: make import as absolute path import

File: lib/src/helpers/validate_tree_solver_min_samples_count.dart [hunk 3]
```diff
@@ -0,0 +1,6 @@
+void validateTreeSolversMinSamplesCount(int minSamplesCount) {
+  if (minSamplesCount < 0) {
```
Comment (gyrdym) [inline lib/src/helpers/validate_tree_solver_min_samples_count.dart]: it also should be equal to 0

File: lib/src/helpers/validate_tree_solver_max_depth.dart [hunk 4]
```diff
@@ -0,0 +1,6 @@
+void validateTreeSolverMaxDepth(int maxDepth) {
+  if (maxDepth < 0) {
```
Comment (gyrdym) [inline lib/src/helpers/validate_tree_solver_max_depth.dart]: it also should not be equal to0

File: lib/src/tree_solver/_helpers/create_decision_tree_solver.dart [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    int maxDepth,
+) {
+  final targetIdx = enumerate(samples.header)
+      .firstWhere((indexedName) => indexedName.value == targetName)
+      .index;
+
+  final featuresIndexedSeries = enumerate(samples.series)
+      .where((indexed) => indexed.index != targetIdx);
+
+  final colIdxToUniqueValues = Map.fromEntries(
```
Comment (gyrdym) [inline lib/src/tree_solver/_helpers/create_decision_tree_solver.dart]: `featureIdxToItsUniqueValues`

File: lib/src/tree_solver/_helpers/create_decision_tree_solver.dart [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+          .toList(growable: false),
+        ),
+      ),
+  );
+
+  final solverFactory = dependencies.getDependency<TreeSolverFactory>();
+
+  return solverFactory.createByType(
+    TreeSolverType.decision,
+    samples.toMatrix(),
```
Comment (gyrdym) [inline lib/src/tree_solver/_helpers/create_decision_tree_solver.dart]: pass `dtype` parameter. By the way, check all the `toMatrix(...)` calls

File: lib/src/tree_solver/leaf_detector/leaf_detector_factory_impl.dart [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+      num minErrorOnNode,
+      int minSamplesCount,
+      int maxDepth,
+  ) {
+    final assessor = _splitAssessorFactory.createByType(assessorType);
+
+    return TreeLeafDetectorImpl(assessor, minErrorOnNode,
+        minSamplesCount, maxDepth);
+  }
+}
```
Comment (gyrdym) [inline lib/src/tree_solver/leaf_detector/leaf_detector_factory_impl.dart]: add new line

File: lib/src/tree_solver/leaf_label/leaf_label.dart [hunk 8]
```diff
@@ -0,0 +1,19 @@
+import 'package:ml_algo/src/common/serializable/serializable_mixin.dart';
+
+class TreeLeafLabel with SerializableMixin {
+  TreeLeafLabel(this.value, {this.probability}) {
+    if (probability != null && (probability < 0 || probability > 1)) {
+      throw Exception('Probability value should be within the range 0..1 '
```
Comment (gyrdym) [inline lib/src/tree_solver/leaf_label/leaf_label.dart]: extract to validator helper

File: lib/src/tree_solver/splitter/nominal_splitter/nominal_splitter_factory_impl.dart [hunk 9]
```diff
@@ -0,0 +1,13 @@
...
+
+class NominalTreeSplitterFactoryImpl implements
+    NominalTreeSplitterFactory {
+
+  const NominalTreeSplitterFactoryImpl();
+
+  @override
+  NominalTreeSplitter create() =>
+      const NominalTreeSplitterImpl();
+}
```
Comment (gyrdym) [inline lib/src/tree_solver/splitter/nominal_splitter/nominal_splitter_factory_impl.dart]: add new line

File: lib/src/tree_solver/splitter/splitter_factory_impl.dart [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+    final numericalSplitter = _numericalSplitterFactory.create();
+    final nominalSplitter = _nominalSplitterFactory.create();
+
+    switch (type) {
+      case TreeSplitterType.greedy:
+        return GreedyTreeSplitter(assessor, numericalSplitter,
+            nominalSplitter);
+
+      default:
+        throw UnsupportedError('Decision tree splitter type $type is not '
```
Comment (gyrdym) [inline lib/src/tree_solver/splitter/splitter_factory_impl.dart]: remove word `Decision`

File: lib/src/tree_solver/splitter/splitter_factory_impl.dart [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+            'supported');
+    }
+  }
+}
```
Comment (gyrdym) [inline lib/src/tree_solver/splitter/splitter_factory_impl.dart]: add new line

File: lib/src/tree_solver/tree_solver_factory.dart [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+import 'package:ml_algo/src/tree_solver/tree_solver_type.dart';
+import 'package:ml_linalg/matrix.dart';
+
+abstract class TreeSolverFactory {
+  TreeSolver createByType(
+      TreeSolverType type,
+      Matrix samples,
+      Iterable<int> featureIndices,
+      int targetIdx,
+      Map<int, List<num>> featureToUniqueValues,
```
Comment (gyrdym) [inline lib/src/tree_solver/tree_solver_factory.dart]: rename to `featureIdxToUniqueValue`

File: lib/src/tree_solver/tree_solver_factory_impl.dart [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+  final TreeLeafLabelFactoryFactory _leafLabelFactoryFactory;
+  final TreeSplitSelectorFactory _splitSelectorFactory;
+
+  @override
+  TreeSolver createByType(
+      TreeSolverType type,
+      Matrix samples,
+      Iterable<int> featureIndices,
+      int targetIdx,
+      Map<int, List<num>> featureToUniqueValues,
```
Comment (gyrdym) [inline lib/src/tree_solver/tree_solver_factory_impl.dart]: do not forget to rename

File: test/classifier/decision_tree_classifier/decision_tree_classifier_test.dart [hunk 14]
```diff
@@ -24,42 +24,94 @@ void main() {
...
+
+    test('should allow minimal error to be 1', () {
+      final actual = DecisionTreeClassifier(dataFrame, 'col_8',
+          minError: 1, minSamplesCount: 1, maxDepth: 3);
+
+      expect(actual, isA<DecisionTreeClassifierImpl>());
+    });
+
+    test('should throw an exception if minimal samples count on node is less '
+        'than zero', () {
```
Comment (gyrdym) [inline test/classifier/decision_tree_classifier/decision_tree_classifier_test.dart]: and equal to 0

File: test/classifier/decision_tree_classifier/decision_tree_classifier_test.dart [hunk 15]
```diff
@@ -24,42 +24,94 @@ void main() {
...
+      final actual = () => DecisionTreeClassifier(dataFrame, 'col_8',
+          minError: 0.5, minSamplesCount: -1, maxDepth: 3);
+
+      expect(actual, throwsException);
+    });
+
+    test('should throw an exception if maximal tree depth value is less '
```
Comment (gyrdym) [inline test/classifier/decision_tree_classifier/decision_tree_classifier_test.dart]: and equal to zero

File: test/tree_solver/leaf_detector/leaf_detector_factory_impl_test.dart [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+        assessorType,
+        minErrorOnNode,
+        minSamplesCount,
+        maxDepth,
+      );
+
+      expect(detector, isA<TreeLeafDetectorImpl>());
+    });
+
+    test('should call split split_assessor factory while creating the instance', () {
```
Comment (gyrdym) [inline test/tree_solver/leaf_detector/leaf_detector_factory_impl_test.dart]: change `split_assessor` with `assessor`

File: test/tree_solver/leaf_label/leaf_label_test.dart [hunk 17]
```diff
@@ -0,0 +1,61 @@
...
+  group('TreeLeafLabel', () {
+    test('should throw an exception if probability is less than 0', () {
+      final labelValue = 5;
+      final probability = -0.001;
+
+      expect(() => TreeLeafLabel(labelValue, probability: probability),
+          throwsException);
+    });
+
+    test('should throw an exception if probability is greater than 1', () {
```
Comment (gyrdym) [inline test/tree_solver/leaf_label/leaf_label_test.dart]: check also edge cases when probability is either 0 or 1

File: test/tree_solver/test_utils.dart [hunk 18]
```diff
@@ -28,8 +28,8 @@ void testTreeNode(
   });
 }
 
-void testLeafLabel(DecisionTreeLeafLabel label,
-    DecisionTreeLeafLabel expectedLabel) {
+void testLeafLabel(TreeLeafLabel label,
+    TreeLeafLabel expectedLabel) {
   expect(label.value, equals(expectedLabel.value));
   expect(label.probability, equals(expectedLabel.probability));
 }
```
Comment (gyrdym) [inline test/tree_solver/test_utils.dart]: add new line

File: test/tree_solver/tree_solver_factory_impl_test.dart [hunk 19]
```diff
@@ -0,0 +1,115 @@
...
+        splitSelectorType,
+        splitAssessorType,
+        splitterType,
+      );
+    });
+
+    tearDown(() {
+      reset(leafDetectorFactoryMock);
+      reset(leafLabelFactoryFactoryMock);
+      reset(splitSelectorFactoryMock);
```
Comment (gyrdym) [inline test/tree_solver/tree_solver_factory_impl_test.dart]: reset also instance mocks
